// karma-qunit/lib/index.js
// ***************************************************************************
// Copyright (c) 2017, Industrial Logic, Inc., All Rights Reserved.
//
// This code is the exclusive property of Industrial Logic, Inc. It may ONLY be
// used by students during Industrial Logic's workshops or by individuals
// who are being coached by Industrial Logic on a project.
//
// This code may NOT be copied or used for any other purpose without the prior
// written consent of Industrial Logic, Inc.
// ****************************************************************************

var path = require('path')

function createPattern (pattern) {
  return {
    pattern: pattern,
    included: true,
    served: true,
    watched: false
  }
}

function initQUnit (files, qunit) {
  files.unshift(createPattern(path.join(__dirname, 'adapter.js')))
  files.unshift(createPattern(require.resolve('qunitjs')))
  if (qunit && qunit.showUI) {
    files.unshift(createPattern(require.resolve('qunitjs/qunit/qunit.css')))
  }
}

initQUnit.$inject = ['config.files', 'config.client.qunit']

module.exports = {
  'framework:qunit': ['factory', initQUnit]
}
