// karma/lib/browser_result.js
// ***************************************************************************
// Copyright (c) 2017, Industrial Logic, Inc., All Rights Reserved.
//
// This code is the exclusive property of Industrial Logic, Inc. It may ONLY be
// used by students during Industrial Logic's workshops or by individuals
// who are being coached by Industrial Logic on a project.
//
// This code may NOT be copied or used for any other purpose without the prior
// written consent of Industrial Logic, Inc.
// ****************************************************************************

var Result = function () {
  var startTime = Date.now()

  this.total = this.skipped = this.failed = this.success = 0
  this.netTime = this.totalTime = 0
  this.disconnected = this.error = false

  this.totalTimeEnd = function () {
    this.totalTime = Date.now() - startTime
  }

  this.add = function (result) {
    if (result.skipped) {
      this.skipped++
    } else if (result.success) {
      this.success++
    } else {
      this.failed++
    }

    this.netTime += result.time
  }
}

module.exports = Result
