// karma/lib/reporters/base_color.js
// ***************************************************************************
// Copyright (c) 2017, Industrial Logic, Inc., All Rights Reserved.
//
// This code is the exclusive property of Industrial Logic, Inc. It may ONLY be
// used by students during Industrial Logic's workshops or by individuals
// who are being coached by Industrial Logic on a project.
//
// This code may NOT be copied or used for any other purpose without the prior
// written consent of Industrial Logic, Inc.
// ****************************************************************************

require('colors')

var BaseColorReporter = function () {
  this.USE_COLORS = true

  this.LOG_SINGLE_BROWSER = '%s: ' + '%s'.cyan + '\n'
  this.LOG_MULTI_BROWSER = '%s %s: ' + '%s'.cyan + '\n'

  this.SPEC_FAILURE = '%s %s FAILED'.red + '\n'
  this.SPEC_SLOW = '%s SLOW %s: %s'.yellow + '\n'
  this.ERROR = '%s ERROR'.red + '\n'

  this.FINISHED_ERROR = ' ERROR'.red
  this.FINISHED_SUCCESS = ' SUCCESS'.green
  this.FINISHED_DISCONNECTED = ' DISCONNECTED'.red

  this.X_FAILED = ' (%d FAILED)'.red

  this.TOTAL_SUCCESS = 'TOTAL: %d SUCCESS'.green + '\n'
  this.TOTAL_FAILED = 'TOTAL: %d FAILED, %d SUCCESS'.red + '\n'
}

// PUBLISH
module.exports = BaseColorReporter
