// karma/lib/reporters/dots.js
// ***************************************************************************
// Copyright (c) 2017, Industrial Logic, Inc., All Rights Reserved.
//
// This code is the exclusive property of Industrial Logic, Inc. It may ONLY be
// used by students during Industrial Logic's workshops or by individuals
// who are being coached by Industrial Logic on a project.
//
// This code may NOT be copied or used for any other purpose without the prior
// written consent of Industrial Logic, Inc.
// ****************************************************************************

var BaseReporter = require('./base')

var DotsReporter = function (formatError, reportSlow, useColors, browserConsoleLogOptions) {
  BaseReporter.call(this, formatError, reportSlow, useColors, browserConsoleLogOptions)

  var DOTS_WRAP = 80
  this.EXCLUSIVELY_USE_COLORS = false
  this.onRunStart = function () {
    this._browsers = []
    this._dotsCount = 0
  }

  this.onBrowserStart = function (browser) {
    this._browsers.push(browser)
  }

  this.writeCommonMsg = function (msg) {
    if (this._dotsCount) {
      this._dotsCount = 0
      msg = '\n' + msg
    }

    this.write(msg)
  }

  this.specSuccess = function () {
    this._dotsCount = (this._dotsCount + 1) % DOTS_WRAP
    this.write(this._dotsCount ? '.' : '.\n')
  }

  this.onBrowserComplete = function (browser) {
    this.writeCommonMsg(this.renderBrowser(browser) + '\n')
  }

  this.onRunComplete = function (browsers, results) {
    if (browsers.length > 1 && !results.disconnected && !results.error) {
      if (!results.failed) {
        this.write(this.TOTAL_SUCCESS, results.success)
      } else {
        this.write(this.TOTAL_FAILED, results.failed, results.success)
      }
    }
  }
}

// PUBLISH
module.exports = DotsReporter
