// log4js/lib/appenders/mailgun.js
// ***************************************************************************
// Copyright (c) 2017, Industrial Logic, Inc., All Rights Reserved.
//
// This code is the exclusive property of Industrial Logic, Inc. It may ONLY be
// used by students during Industrial Logic's workshops or by individuals
// who are being coached by Industrial Logic on a project.
//
// This code may NOT be copied or used for any other purpose without the prior
// written consent of Industrial Logic, Inc.
// ****************************************************************************

"use strict";
var layouts = require('../layouts');
var layout;
var config;
var mailgun;

function mailgunAppender(_config, _layout) {

    config = _config;
    layout = _layout || layouts.basicLayout;

    return function (loggingEvent) {

        var data = {
            from: _config.from,
            to: _config.to,
            subject: _config.subject,
            text: layout(loggingEvent, config.timezoneOffset)
        };

        mailgun.messages().send(data, function (error, body) {
            if (error !== null) console.error("log4js.mailgunAppender - Error happened", error);
        });
    };
}

function configure(_config) {
    config = _config;

    if (_config.layout) {
        layout = layouts.layout(_config.layout.type, _config.layout);
    }

    mailgun = require('mailgun-js')({
        apiKey: _config.apikey,
        domain: _config.domain
    });

    return mailgunAppender(_config, layout);
}

exports.appender = mailgunAppender;
exports.configure = configure;
