// karma/lib/init/color_schemes.js
// ***************************************************************************
// Copyright (c) 2017, Industrial Logic, Inc., All Rights Reserved.
//
// This code is the exclusive property of Industrial Logic, Inc. It may ONLY be
// used by students during Industrial Logic's workshops or by individuals
// who are being coached by Industrial Logic on a project.
//
// This code may NOT be copied or used for any other purpose without the prior
// written consent of Industrial Logic, Inc.
// ****************************************************************************

var COLORS_ON = {
  RESET: '\x1B[39m',
  ANSWER: '\x1B[36m', // NYAN
  SUCCESS: '\x1B[32m', // GREEN
  QUESTION: '\x1B[1m', // BOLD
  question: function (str) {
    return this.QUESTION + str + '\x1B[22m'
  },
  success: function (str) {
    return this.SUCCESS + str + this.RESET
  }
}

var COLORS_OFF = {
  RESET: '',
  ANSWER: '',
  SUCCESS: '',
  QUESTION: '',
  question: function (str) {
    return str
  },
  success: function (str) {
    return str
  }
}

exports.ON = COLORS_ON
exports.OFF = COLORS_OFF
